crates/ruma-events/src/matrix_rtc/member_event.rs

```diff
@@ -0,0 +1,234 @@
+    /// # Arguments
+    ///
+    /// * `origin_server_ts` - optionally the `origin_server_ts` can be passed as a fallback in the
+    ///   Membership does not contain [`LegacyMembershipData::created_ts`]. (`origin_server_ts` will
+    ///   be ignored if [`LegacyMembershipData::created_ts`] is `Some`)
+    pub fn active_memberships(
+        &self,
+        origin_server_ts: Option<MilliSecondsSinceUnixEpoch>,
+    ) -> Vec<MembershipData<'_>> {
+        self.0.active_memberships(origin_server_ts)
+    }
+    /// All the memberships for this event. Can only contain multiple elements in the case of legacy
+    /// `m.call.member` state events.
+    pub fn memberships(&self) -> Vec<MembershipData<'_>> {
+        self.0.memberships()
+    }
+
+    /// Set the `created_ts` of each [`MembershipData::Legacy`] in this event.
+    ///
+    /// Each call member event contains the `origin_server_ts` and `content.create_ts`.
+    /// `content.create_ts` is undefined for the initial event of a session (because the
+    /// `origin_server_ts` is not known on the client).
+    /// In the rust sdk we want to copy over the `origin_server_ts` of the event into the
+    /// (This allows to use `MinimalStateEvents` and still be able to determine if a
+    /// expired)
+    pub fn set_created_ts_if_none(&mut self, origin_server_ts: MilliSecondsSinceUnixEpoch) {
+        self.0.set_created_ts_if_none(origin_server_ts);
+    }
+}
+
+/// The PossiblyRedacted version of [`CallMemberEventContent`].
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[allow(clippy::exhaustive_structs)]
+pub struct PossiblyRedactedCallMemberEventContent(pub MemberEventContent);
+
+impl ruma_events::content::EventContent for PossiblyRedactedCallMemberEventContent {
+    type EventType = StateEventType;
+    fn event_type(&self) -> Self::EventType {
+        StateEventType::CallMember
+    }
+}
+
+impl PossiblyRedactedStateEventContent for PossiblyRedactedCallMemberEventContent {
+    type StateKey = String;
+}
+
+/// The object containing the actual content of a [`CallMemberEventContent`].
+///
+/// There are two versions of [`CallMemberEventContent`]. One for legacy events,
+/// which contain an array of memberships and one version for the new format that just
+/// has one membership.
+///
+/// This is an untagged serde enum to be able to parse both versions.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(untagged)]
+pub enum MemberEventContent {
+    /// The legacy format for m.call.member events. (An array of memberships. The devices of one
+    /// user.)
+    LegacyContent(LegacyMembershipContent),
+    /// Normal membership events. One event per membership. Multiple state keys will
+    /// be used to describe multiple devices for one user.
+    SessionContent(SessionMembershipData),
+    /// An empty content means this user has been in a rtc session but is not anymore.
+    Empty {},
```

Comment:
This happens only when the event is redacted, or is it possible to send an empty event content?

Reply:
Its possible and per spec to send an empty content. This means that the user is currently not connected to the rtc session.

Reply:
Okay, we lack a constructor for it maybe?